### PR TITLE
Llvm support

### PR DESCRIPTION
**Pull request by zhouxinyu0723**
It is just a draft. Still has some errors.

### CODE REVIEW

File: template_project/microtvm_api_server.py [hunk 1]
```diff
@@ -124,6 +124,13 @@ class BoardError(Exception):
 
 
 PROJECT_OPTIONS = [
+    server.ProjectOption(
+        "toolchain",
+        optional=["build", "flash", "open_transport"],
+        type="str",
+        choices=["llvm", "gcc"], # I do not know how to parse it.
```
Comment (zhouxinyu0723) [inline template_project/microtvm_api_server.py:131]: I do not know how to parse it.

File: template_project/cmake/LookupClang.cmake [hunk 2]
```diff
@@ -0,0 +1,50 @@
+set(CLANG_VERSIONS 15 14)
+
+set(LLVM_DIR "" CACHE PATH "Lookup path for the llvm installation.")
+# https://stackoverflow.com/questions/28613394/check-cmake-cache-variable-in-toolchain-file
+# Problem: CMake runs toolchain files multiple times, but can't read cache variables on some runs.
+# Workaround: On first run (in which cache variables are always accessible), set an intermediary environment variable.
```
Comment (zhouxinyu0723) [inline template_project/cmake/LookupClang.cmake:6]: A workaround in the cmake tool chain related cmake file.

Comment (PhilippvK) [reply template_project/cmake/LookupClang.cmake:6]: I will look into this.

**Comment by zhouxinyu0723** [comment]
micro_autotune_gvsoc_tflite llvm linking problem. 

[examplesmicro_autotune_gvsoc_tflite gcc.txt](https://github.com/PhilippvK/microtvm-gvsoc-template/files/10753771/examplesmicro_autotune_gvsoc_tflite.gcc.txt)
[examplesmicro_autotune_gvsoc_tflite llvm.txt](https://github.com/PhilippvK/microtvm-gvsoc-template/files/10753772/examplesmicro_autotune_gvsoc_tflite.llvm.txt)

**Comment by PhilippvK** [comment]
@zhouxinyu0723 Those missing symbols can be found the the C math library. Can you try just adding `-lm` to the linker flags?

**Comment by zhouxinyu0723** [comment]
> 

Yes, it solves the problem.

**Comment by PhilippvK** [comment]
I am running in this problem:

```
CMake Error at cmake/LookupClang.cmake:38 (MESSAGE):
  Unable to find lld executable
Call Stack (most recent call first):
  cmake/LookupClang.cmake:46 (do_lookup)
  cmake/toolchain_llvm.cmake:3 (INCLUDE)
  /nas/ei/share/TUEIEDAprojects/SystemDesign/tools/cmake/3.22.2/share/cmake-3.22/Modules/CMakeDetermineSystem.cmake:124 (include)
  CMakeLists.txt:68 (project)
```

**Comment by PhilippvK** [comment]
which llvm toolchain are you using?

(it failed to link with my custom build pulp llvm, however default llvm 14 seems to work fine with std extensions)

**Comment by zhouxinyu0723** [comment]
> which llvm toolchain are you using?
> 
> (it failed to link with my custom build pulp llvm, however default llvm 14 seems to work fine with std extensions)

If I understand correctly, the custom build does not have lld. So which release version pulp-llvm should I use?

**Comment by PhilippvK** [comment]
Here is the requyired patch:

```
diff --git a/examples/micro_autotune_gvsoc.py b/examples/micro_autotune_gvsoc.py
index 7a38700..9965f6a 100644
--- a/examples/micro_autotune_gvsoc.py
+++ b/examples/micro_autotune_gvsoc.py
@@ -61,7 +61,7 @@ project_options = {
     "pulp_llvm_path": PULP_LLVM_DIR,
     "toolchain": "llvm", # llvm for compilation with llvm, gcc for complition with gcc
     "arch": "rv32imac",
-    "abi": "ilp64",
+    "abi": "ilp32",
     "memory_size_bytes": 2**17,
 }

@@ -136,7 +136,7 @@ measure_option = tvm.autotvm.measure_option(builder=builder, runner=runner)
 # Run Autotuning
 ################

-num_trials = 2
+num_trials = 20
 for i, task in enumerate(tasks):
     prefix = "[Task %2d/%2d] " % (i + 1, len(tasks))
     tuner = tvm.autotvm.tuner.GATuner(task)
diff --git a/template_project/cmake/toolchain_llvm.cmake b/template_project/cmake/toolchain_llvm.cmake
index 74b8c84..4802582 100644
--- a/template_project/cmake/toolchain_llvm.cmake
+++ b/template_project/cmake/toolchain_llvm.cmake
@@ -48,4 +48,6 @@ SET(CV_LDFLAGS "\
 SET(CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} ${CV_CFLAGS} ${CV_CPPFLAGS} -DSTDIO_FAKE=2 -DSTDIO_UART=1 -DSTDIO_NULL=0 -DCONFIG_STDIO=2 ")
 SET(CMAKE_C_FLAGS  "${CMAKE_C_FLAGS} ${CV_CFLAGS} ${CV_CPPFLAGS} -DSTDIO_FAKE=2 -DSTDIO_UART=1 -DSTDIO_NULL=0 -DCONFIG_STDIO=2 ")
 SET(CMAKE_ASM_FLAGS "${CMAKE_ASM_FLAGS} ${CV_ASFLAGS} ${CV_CPPFLAGS} -DportasmHANDLE_INTERRUPT=vSystemIrqHandler ")
-SET(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -DportasmHANDLE_INTERRUPT=vSystemIrqHandler ${CV_CFLAGS} -g3 ${CV_LDFLAGS}")
\ No newline at end of file
+SET(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -DportasmHANDLE_INTERRUPT=vSystemIrqHandler ${CV_CFLAGS} ${CV_LDFLAGS}")
+
+SET(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fuse-ld=lld")
```

**Comment by PhilippvK** [comment]
I used the following llvm: https://versaweb.dl.sourceforge.net/project/pulp-llvm-project/nightly/riscv32-pulp-llvm-ubuntu2004.tar.gz

**Comment by PhilippvK** [comment]
@zhouxinyu0723 thank you for the fixes. Can I merge this now?

**Comment by zhouxinyu0723** [comment]
Yes

**Pull request merged**